poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+ *   * VerifyPromise
+ *   * VerifyShare
+ *   * Sign
+ *   * RevealShare
+ *   * Blame
+ *
+ * - Clients
+ *   * VerifyPromise
+ *   * VerifyRevealedShare
+ *
+ * - All
+ *   * UnmarshalInit
+ *   * VerifySignature
+ *   * VerifyBlame
+ */
+type Promise struct {
+
+	// The id of the promise used to differentiate it from others
+	// The id is the short term public key of the private key being promised
+	id string
+
+	// The cryptographic suite to use for the shared secrets
+	shareSuite abstract.Suite
+
+	// The suite used for the long term keys stored in the Promise
+	keySuite abstract.Suite
+
+	// The minimum number of shares needed to reconstruct the secret
+	t int
+
+	// The minimum number of PromiseSignatures approving the Promise that
+	// are needed before the Promise is certified. t <= r <= n
+	r int
+
+	// The total number of shares
+	n int
+
+	// The long-term public key of the promiser
+	pubKey abstract.Point
+
+	// The public polynomial that is used to verify the shared secrets
+	pubPoly poly.PubPoly
+
+	// A list of servers who will act as insurers of the Promise. The list
+	// contains the long-term public keys of the insurers
+	insurers []abstract.Point
+
+	// The list of shared secrets to be sent to the insurers. They are
+	// encrypted with Diffie-Hellman shared secrets between the insurer
+	// and the promiser.
+	secrets []abstract.Secret
+}
+
+/* Constructs a new Promise to guarentee a secret.
+ *
+ * Arguments
+ *    secretPair   = the keypair of the secret to be promised
+ *    longPair     = the long term keypair of the promiser
+ *    t            = minimum number of shares needed to reconstruct the secret.
+ *    r            = minimum PromiseSignatures needed to certify the Promise
+ *    insurers     = a list of the long-term public keys of the insurers.
+ *
+ * Returns
+ *   A newly constructed Promise
+ */
+func (p *Promise) ConstructPromise(secretPair *config.KeyPair,
+	longPair *config.KeyPair, t, r int, insurers []abstract.Point) *Promise {
+	p.id = secretPair.Public.String()
+	p.t = t
+	p.r = r
+	p.n = len(insurers)
+	p.shareSuite = secretPair.Suite
+	p.keySuite = longPair.Suite
+	p.pubKey = longPair.Public
+	p.insurers = insurers
+	p.secrets = make([]abstract.Secret, p.n, p.n)
+
+	// Verify that t <= r <= n
+	if p.n < p.t {
+		panic("Not enough insurers for the secret")
+	}
+	if p.r < p.t {
+		p.r = p.t
+	}
+	if p.r > p.n {
+		p.r = p.n
+	}
+
+	// Create the public polynomial and private shares. The number of shares
+	// should be equal to the number of insurers.
+	pripoly := new(poly.PriPoly).Pick(p.shareSuite, p.t,
+		secretPair.Secret, random.Stream)
+	prishares := new(poly.PriShares).Split(pripoly, p.n)
+	p.pubPoly = poly.PubPoly{}
+	p.pubPoly.Commit(pripoly, nil)
+
+	// Populate the secrets array with the shares encrypted by a Diffie-
+	// Hellman shared secret between the promiser and appropriate insurer
+	for i := 0; i < p.n; i++ {
+		diffieBase := p.keySuite.Point().Mul(insurers[i], longPair.Secret)
```

Comment:
Here is (I think) why there can only be one suite: longPair.Secret gets used both in secret-sharing and in DH key agreement.